crates/tests/src/sync.rs

```diff
@@ -0,0 +1,227 @@
+
+        // geth starts in dev mode, we can spawn it, mine blocks, and shut it down we need to clone
+        // it because we will be reusing the geth config when we restart p2p
+        let mut instance = geth.spawn();
+
+        // take the stderr of the geth instance and print it
+        let stderr = instance.stderr().unwrap();
+
+        // print logs in a new task
+        task::spawn(async move {
+            let mut err_reader = BufReader::new(stderr);
+
+            loop {
+
+                let mut buf = String::new();
+                if let Ok(line) = err_reader.read_line(&mut buf) {
+                    if line == 0 {
+                        tokio::time::sleep(Duration::from_nanos(1)).await;
+                        continue
+                    }
+                    dbg!(buf);
+                }
+            }
+        });
+
+        // === check that we have the same genesis hash ===
+
+        // set up provider
+        let provider = produce_blocks(instance.port(), signing_key).await;
+
+        // get genesis hash
+        let genesis_block =
+            provider.get_block(0).await.unwrap().expect("a genesis block should exist");
+
+        // get our hash
+        let sealed_genesis = genesis_header(&genesis.clone()).seal();
+
+        // let's just convert into a reth header and compare
+        let geth_genesis_header = block_to_header(genesis_block.clone()).seal();
+        assert_eq!(geth_genesis_header, sealed_genesis, "genesis headers should match, we computed {sealed_genesis:#?} but geth computed {geth_genesis_header:#?}");
+
+        // make sure we have the same genesis hash
+        let genesis_hash: H256 = genesis_block.hash.unwrap().0.into();
+        let sealed_hash = sealed_genesis.hash();
+        assert_eq!(sealed_hash, genesis_hash, "genesis hashes should match, we computed {sealed_hash:?} but geth computed {genesis_hash:?}");
+
+        // === create many blocks ===
+
+        let nonces = 0..10000u64;
+        let txs = nonces
+            .map(|nonce| {
+                // create a tx that just sends to the zero addr
+                TypedTransaction::Eip1559(Eip1559TransactionRequest::new()
+                    .to(H160::zero())
+                    .value(1u64)
+                    .nonce(nonce))
+            });
+
+        for tx in txs {
+            // send the tx - geth will mine a block with just this transaction in it if we await
+            // here rather than joining concurrent sends
+            provider.send_transaction(tx, None).await.unwrap();
+        }
+
+        // wait for a certain number of blocks to be mined
+        let block = provider.get_block_number().await.unwrap();
+        println!("block num after creating transactions: {block}");
+        assert!(block > U64::zero());
+
+        // get the tip so we can send it to reth
+        let tip_block = provider.get_block(BlockNumber::Latest).await.unwrap().unwrap();
+        let tip_hash = tip_block.hash.unwrap().0.into();
```

Comment:
All of this setup code should be  on a helper, probably on the test clique geth struct